### ISSUE TITLE
Tiny fraction of reads mapped

### ISSUE THREAD

**Issue by Blahah** [amplab/snap#61]
via @ctb

User input 114396588 pairs, of which only 77778 were reported in the BAM file.

SNAP logs:

```
Loading index from directory... 0s.  236823466 bases, seed size 23
Aligning.
Welcome to SNAP version 1.0beta.18.

sched_setaffinity: Invalid argument
sched_setaffinity: Invalid argument
sched_setaffinity: Invalid argument
sched_setaffinity: Invalid argument
```

Could those log messages have something to do with it? Can provide input data if necessary.

**Comment by ctb** [comment]
On Sun, Nov 01, 2015 at 09:21:43AM -0800, Richard Smith-Unna wrote:
> via @ctb
> 
> User input 114396588 pairs, of which only 77778 were reported in the BAM file.
> 
> SNAP logs:
> 
> ```
> Loading index from directory... 0s.  236823466 bases, seed size 23
> Aligning.
> Welcome to SNAP version 1.0beta.18.
> 
> sched_setaffinity: Invalid argument
> sched_setaffinity: Invalid argument
> sched_setaffinity: Invalid argument
> sched_setaffinity: Invalid argument
> ```
> 
> Could those log messages have something to do with it? Can provide input data if necessary.

Maybe, although the advice on line is to ignore it.

I recompiled snap and am now trying to figure out how to get transrate
to find it.

**Comment by ctb** [comment]
Same problem as #60.

**Comment by bolosky** [comment]
No, those messages don’t have anything to do with read mapping.

SNAP tries to bind the aligner threads to cores, which somewhat improves efficiency because the hardware doesn’t have to move the cache state to follow the thread.  This message means that it failed to bind a thread to a core, which usually happens when you give it –t with more threads than there are cores in the system.  When that happens, the extra threads float to whatever core is idle, which might affect performance but won’t affect behavior.

Seeing lots of reads in the input which don’t make it into the output is probably because of one of two things.  One is that the paired read matcher can’t match ends of reads to one another, either because RNEXT and PNEXT aren’t filled in, or because of a bug that Ravi’s working on now; it usually generates a message at the end of the alignment to this effect.  The other reason is that the input reads are marked with the secondary or supplementary alignment flags (0x100 and 0x800), which ordinarily are dropped during the input phase because these aren’t real reads from the sequencer, they’re artifacts produced by a previous aligner.  If you want to keep them, you can say –sa.

If it’s not either of those things, please let me know and I’ll try to figure out what’s going on.

--Bill

**Comment by ctb** [comment]
On Sun, Nov 01, 2015 at 10:11:05AM -0800, Bill Bolosky wrote:
> No, those messages don???t have anything to do with read mapping.

Yes, sorry, the original problem was few reads mapping, and that was the only
message in the log file; but it's not the problem.

**Comment by Blahah** [comment]
These are FASTQ format input, so I don't think it can be `RNEXT/PNEXT` or supplementary aln flags.

**Comment by bolosky** [comment]
Well then that’s very strange.

By “reported in the BAM file” you mean that are there at all, not there and mapped?  That is, you’re saying that it’s completely losing reads rather than simply failing to map them, right?  What does SNAP print out at the end of its run when it reports read counts (the line that starts “Total Reads   Aligned MAPQ >= 10…”?

--Bill

**Comment by ctb** [comment]
tl; dr? I can chase down the error messages if you really want, but I think the root problem is #60.

Longer version:

I was using transrate, and getting very low mapping stats. The version of transrate that I was using comes with snap 1.0b18, which had the above error message. @Blahah thought it might be the problem behind the low mapping rate, so he created this issue.

In the meantime, I compiled my own version of snap-aligner, identified as 1.0b20 (<= latest from github) and figured out that (variously) 1.0b20 crashed on 'snap-aligner paired' when used with the index that I'd created, and that 1.0b18 misbehaved in some way with the same index.  At some point some snap-aligner command said, hey, I hate your FASTA header format, and so I shortened the headers in my reference transcriptome. transrate out-of-the-box (with snap 1.0b18) now maps a decent number of reads  and all is well.

If you want error messages or verification, I am happy to provide them, but I suspect the root cause is the length of my FASTA headers, which is documented in #60. If you fix that I can re-run everything with the original data and verify that it solves all the problems!

**Comment by bolosky** [comment]
OK. I have a fix for that that’s almost ready to go.  I’ll try to get it checked in tomorrow.

**Comment by ctb** [comment]
great!

**Comment by bolosky** [comment]
I pushed a fix for very long contig names in beta.21 (and dev.91). You should try that and see if it helps.
